imports/plugins/core/payments/server/no-meteor/mutations/approveOrderPayments.test.js

```diff
@@ -0,0 +1,120 @@
+import mockContext from "/imports/test-utils/helpers/mockContext";
+import approveOrderPayments from "./approveOrderPayments";
+
+beforeEach(() => {
+  jest.resetAllMocks();
+});
+
+test("throws if orderId isn't supplied", async () => {
+  await expect(approveOrderPayments(mockContext, {
+    paymentIds: ["1"],
+    shopId: "SHOP_ID"
+  })).rejects.toThrowErrorMatchingSnapshot();
+});
+
+test("throws if paymentIds isn't supplied", async () => {
+  await expect(approveOrderPayments(mockContext, {
+    orderId: "abc",
+    shopId: "SHOP_ID"
+  })).rejects.toThrowErrorMatchingSnapshot();
+});
+
+test("throws if shopId isn't supplied", async () => {
+  await expect(approveOrderPayments(mockContext, {
+    orderId: "abc",
+    paymentIds: ["1"]
+  })).rejects.toThrowErrorMatchingSnapshot();
+});
+
+test("throws if the order doesn't exist", async () => {
+  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(null));
+
+  await expect(approveOrderPayments(mockContext, {
+    orderId: "abc",
+    paymentIds: ["1"],
+    shopId: "SHOP_ID"
+  })).rejects.toThrowErrorMatchingSnapshot();
+});
+
+test("throws if permission check fails", async () => {
+  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
+    _id: "abc",
+    payments: [
+      {
+        _id: "1",
+        status: "created"
+      }
+    ],
+    shopId: "SHOP_ID"
+  }));
+
+  mockContext.userHasPermission.mockReturnValueOnce(false);
+
+  await expect(approveOrderPayments(mockContext, {
+    orderId: "abc",
+    paymentIds: ["1"],
+    shopId: "SHOP_ID"
+  })).rejects.toThrowErrorMatchingSnapshot();
+
+  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders"], "SHOP_ID");
+});
+
+test("throws if the order payment doesn't exist", async () => {
+  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
+    payments: [
+      {
+        _id: "2"
+      }
+    ],
+    shopId: "SHOP_ID"
+  }));
+
+  mockContext.userHasPermission.mockReturnValueOnce(true);
+
+  await expect(approveOrderPayments(mockContext, {
+    orderId: "abc",
+    paymentIds: ["1"],
+    shopId: "SHOP_ID"
+  })).rejects.toThrowErrorMatchingSnapshot();
+});
+
+test("updates an order status", async () => {
+  mockContext.userHasPermission.mockReturnValueOnce(true);
+
+  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
+    _id: "abc",
+    payments: [
+      {
+        _id: "1",
+        status: "created"
+      }
+    ],
+    shopId: "SHOP_ID"
+  }));
+
+  await approveOrderPayments(mockContext, {
+    orderId: "abc",
+    paymentIds: ["1"],
+    shopId: "SHOP_ID"
+  });
+
+  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
+    modifiedCount: 1,
+    value: {}
```

Comment:
@aldeed can you help me figure out why this test is failing?

I get the error:

```
TypeError: Cannot destructure property `value` of 'undefined' or 'null'.

      65 |
      66 |   return { order: updatedOrder };
    > 67 | }
         |  ^
      68 |

      at _approveOrderPayments (imports/plugins/core/payments/server/no-meteor/mutations/approveOrderPayments.js:67:2)
```

Which makes me think my mock isn't returning correctly? But if I'm just checkout what the `findOneAndUpdate` was called with, does that matter? I've followed the pattern from a lot of other tests here, but can't get it working.

Reply:
`findOneAndUpdate.mockReturnValueOnce` is after the `await approveOrderPayments` call, so it's mocked too late. Move it to right after the `findOne.mockReturnValueOnce`.

Reply:
So simple, thanks!